python_modules/libraries/dagster-dbt/dagster_dbt/freshness_builder.py

```diff
@@ -0,0 +1,122 @@
+import datetime
+from typing import Sequence
+
+from dagster import (
+    AssetsDefinition,
+)
+from dagster._core.definitions.asset_check_factories.freshness_checks.last_update import (
+    build_last_update_freshness_checks,
+)
+from dagster._core.definitions.asset_check_factories.freshness_checks.time_partition import (
+    build_time_partition_freshness_checks,
+)
+from dagster._core.definitions.asset_check_factories.utils import DEFAULT_FRESHNESS_SEVERITY
+from dagster._core.definitions.asset_check_spec import AssetCheckSeverity
+from dagster._core.definitions.asset_checks import AssetChecksDefinition
+from dagster._core.definitions.time_window_partitions import TimeWindowPartitionsDefinition
+from dagster._core.errors import DagsterInvariantViolationError
+
+from dagster_dbt.utils import ASSET_RESOURCE_TYPES
+
+from .asset_utils import get_manifest_and_translator_from_dbt_assets
+
+
+def build_freshness_checks_from_dbt_assets(
+    dbt_assets: Sequence[AssetsDefinition],
+) -> Sequence[AssetChecksDefinition]:
```

Comment:
I'd be wary of accepting a sequence of `dbt_assets` here.

If a user has `n` `dbt_assets`, we'd potentially define the same asset check definitions `n` times.

We should:
- enforce that `dbt_assets` is singular
- let this optionally take in a manifest and `dagster_dbt_translator` instead.

Reply:
Do we have an easy way for users to retrieve the manifest and translator / are those objects readily available? I would like to avoid the situation where there are two sets of valid parameters, and all parameters are optional on the signature (high parameter anxiety), so if we feel like the manifest and translator are more scalable params, then let's just go with those. @sryza do you have any thoughts to the contrary here?

Reply:
> I understand why we would be defining the same asset check definitions n times - is the concern that users will pass in the same AssetsDefinition object n times? Or am I missing something.

The current implementation reads the manifest from the `@dbt_assets` definition and constructs a freshness check for each `dbt_resource_props` in the manifest.

If you pass in two different `@dbt_assets` definitions that work off of the same manifest, then you would have created the same freshness checks twice.

Reply:
right - I'm saying once that problem is solved / we're operating on the list of asset keys directly, we should be able to pass a sequence, or is there anything else?

Reply:
Sure, let's still pass in a sequence of definitions so that this API is inline with the other signatures.

We should de-dup the parameters passed in `Definitions` as a follow up.